### PR TITLE
feat: adding cache functionality in Receiver for unsorted datapoints

### PR DESCRIPTION
**Pull request by wei3erHase**
This PR aims to solve the issue of data arriving in unsorted order, which made the datapoint ignored, and all the subsequent also invalid, here's what was happening:
- DataFeed bridges nonce 1,2,3,4
- The Bridge process it unsorted, and txs are executed in order 1,3,2,4
- Nonce 1 arrives ✅
- Nonce 3 arrives ❌ (missing nonce 2)
- Nonce 2 arrives ✅ 
- Nonce 4 arrives ❌ (missing nonce 3)
- Nonce N arrives ❌ (bricked until nonce 3 is re-broadcast)

The feature includes a mechanism, in which when a nonce is rejected, it will store it on contract, waiting for the correct nonce to arrive, in the example above:
- Nonce 1 arrives ✅
- Nonce 3 arrives ❌ > stores it on cache
- Nonce 2 arrives ✅ 
- Nonce 4 arrives ❌ > reads cache > writes nonce 3 ✅ > writes nonce 4 ✅ 

The PR also deprecates the observation data from `ObservationsAdded` event (not needed, info is already emitted in L1, and we can use `(pool,nonce)` as observation id) and adds a `ObservationsCached` event for the cases where the observation is invalid

TODO:
- [ ] fix tests

### CODE REVIEW

File: test/e2e/oracle-sidechain.spec.ts [hunk 1]
```diff
@@ -90,5 +90,25 @@ describe('@skip-on-coverage OracleSidechain.sol', () => {
...
+      const tx3 = await allowedDataReceiver.internalAddObservations(obs2, salt, 2); // ordered (after 1)
+      const tx4 = await allowedDataReceiver.internalAddObservations(obs4, salt, 4); // should include 3
+
+      // NOTE: chai is failing to compare the emitted struct[], so we are not checking the arguments
+      await expect(tx1).to.emit(allowedDataReceiver, 'ObservationsAdded').withArgs(salt, 1, deployer.address);
+      await expect(tx2).to.emit(allowedDataReceiver, 'ObservationsCached').withArgs(salt, 3, deployer.address);
+      await expect(tx2).not.to.emit(allowedDataReceiver, 'ObservationsAdded');
+      await expect(tx3).to.emit(allowedDataReceiver, 'ObservationsAdded').withArgs(salt, 2, deployer.address);
+      await expect(tx4).to.emit(allowedDataReceiver, 'ObservationsAdded').withArgs(salt, 3, deployer.address);
+      await expect(tx4).to.emit(allowedDataReceiver, 'ObservationsAdded').withArgs(salt, 4, deployer.address);
```
Comment (0xJabberwock) [inline test/e2e/oracle-sidechain.spec.ts:111]: Why delay in writing `obs3` when it could be done in `tx3`?

Comment (wei3erHase) [reply test/e2e/oracle-sidechain.spec.ts:111]: to show the behaviour of observations arriving disordered

Comment (0xJabberwock) [reply test/e2e/oracle-sidechain.spec.ts:111]: Wouldn't writing `obs3` in `tx3` still show the behavior of disordered observations, as the cacheing happened in `tx2`?

**Comment by wei3erHase** [comment]
Closed in favour of #4 & #5

File: test/e2e/oracle-sidechain.spec.ts (hunk 1 above)
Comment (wei3erHase) [reply test/e2e/oracle-sidechain.spec.ts]: ahá! yes, that was initially just a result of the design, and then it was definitely intended, the alternative is doing `while(true)` and rely on the `break` to exit the loop.

should the caché be fully loaded of new observations (sth that may actually happen!), any write tx would revert with `out of gas`

so, instead, this mechanism adds observations only up to the current nonce being sent, so if there's 1M observations in the cache, we could only process from 0-100, then 100-200,... and the receiver wouldn't be bricked forever

great catch 🙌🏻

**Pull request closed without merging**